operators/pkg/instance-controller/containers_test.go

```diff
@@ -0,0 +1,325 @@
+	clctx "github.com/netgroup-polito/CrownLabs/operators/pkg/context"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/forge"
+	instance_controller "github.com/netgroup-polito/CrownLabs/operators/pkg/instance-controller"
+)
+
+var _ = Describe("Generation of the container based instances", func() {
+	var (
+		ctx           context.Context
+		clientBuilder fake.ClientBuilder
+		reconciler    instance_controller.InstanceReconciler
+
+		instance    clv1alpha2.Instance
+		environment clv1alpha2.Environment
+
+		objectName types.NamespacedName
+		svc        corev1.Service
+		deploy     appsv1.Deployment
+
+		ownerRef metav1.OwnerReference
+
+		containerOpts forge.ContainerEnvOpts
+
+		err error
+	)
+
+	const (
+		instanceName      = "kubernetes-0000"
+		instanceNamespace = "tenant-tester"
+		templateName      = "kubernetes"
+		templateNamespace = "workspace-netgroup"
+		environmentName   = "control-plane"
+		tenantName        = "tester"
+
+		image       = "internal/registry/image:v1.0"
+		cpu         = 2
+		cpuReserved = 25
+		memory      = "1250M"
+		disk        = "20Gi"
+	)
+
+	BeforeEach(func() {
+		ctx = ctrl.LoggerInto(context.Background(), logr.Discard())
+		clientBuilder = *fake.NewClientBuilder().WithScheme(scheme.Scheme).WithObjects(
+			&clv1alpha2.Template{ObjectMeta: metav1.ObjectMeta{Name: templateName, Namespace: templateNamespace}},
+			&clv1alpha1.Tenant{ObjectMeta: metav1.ObjectMeta{Name: tenantName}},
+		)
+
+		instance = clv1alpha2.Instance{
+			ObjectMeta: metav1.ObjectMeta{Name: instanceName, Namespace: instanceNamespace},
+			Spec: clv1alpha2.InstanceSpec{
+				Running:  true,
+				Template: clv1alpha2.GenericRef{Name: templateName, Namespace: templateNamespace},
+				Tenant:   clv1alpha2.GenericRef{Name: tenantName},
+			},
+		}
+		environment = clv1alpha2.Environment{
+			Name:            environmentName,
+			EnvironmentType: clv1alpha2.ClassContainer,
+			Image:           image,
+			Resources: clv1alpha2.EnvironmentResources{
+				CPU:                   cpu,
+				ReservedCPUPercentage: cpuReserved,
+				Memory:                resource.MustParse(memory),
+				Disk:                  resource.MustParse(disk),
+			},
+		}
+
+		objectName = forge.NamespacedName(&instance)
+
+		svc = corev1.Service{}
+		deploy = appsv1.Deployment{}
+
+		ownerRef = metav1.OwnerReference{
+			APIVersion:         clv1alpha2.GroupVersion.String(),
+			Kind:               "Instance",
+			Name:               instance.GetName(),
+			UID:                instance.GetUID(),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+			Controller:         pointer.BoolPtr(true),
+		}
+	})
+
+	JustBeforeEach(func() {
+		containerOpts = forge.ContainerEnvOpts{
+			ImagesTag:     "v1.2.3",
+			XVncImg:       "x-vnc",
+			WebsockifyImg: "wskfy",
+			MyDriveImg:    "mydrive",
+			MyDriveImgTag: "v4.5.6",
+		}
+
+		client := FakeClientWrapped{Client: clientBuilder.Build()}
+		reconciler = instance_controller.InstanceReconciler{
+			Client: client, Scheme: scheme.Scheme,
```

Comment:
```suggestion
			Client: clientBuilder.Build(), Scheme: scheme.Scheme,
```

The FakeClientWrapper is remained from another test (also in the VM one should not be present)

Reply:
Should I change it in vms test as well then?

Reply:
Yes, thanks